File: Ejercicios/builder.py

class Pizza:
    def __init__(self):
        self.size = None
        self.dough = None
        self.ingredients = []
        self.price = 0
        
    def __str__(self):
        text = f"Pizza: {self.size}, masa {self.dough}\n"
        text += "Ingredientes: "+", ".join(self.ingredients) + "\n"
        text += f"Precio total: ${self.price:}"
        return text
    
class PizzaBuilder:
    def __init__(self):
        self.pizza = Pizza()
        
    def set_size(self, size):
        self.pizza.size = size
        if size == "pequeña":
            self.pizza.price += 5
        elif size == "mediana":
            self.pizza.price += 10
        elif size == "grande":
            self.pizza.price += 15
        return self
    
    def set_dough(self, dough):
        self.pizza.dough = dough
        return self
    
    def add_ingredient(self, ingredient,cost):
        self.pizza.ingredients.append(ingredient)
        self.pizza.price += cost
        return self
    
    def build(self):
        return self.pizza
    

builder = PizzaBuilder()
pizza = (builder.set_size("mediana")
            .set_dough("fina")
            .add_ingredient("queso extra", 2)
            .add_ingredient("pepperoni", 3)
            .build())

print(pizza)

File: Ejercicios/Builder.java
package Ejercicios;
import java.util.ArrayList;
import java.util.List;

class Pizza{
    private String size;
    private String dough;
    private List<String> ingredients = new ArrayList<>();
    private double price = 0;

    public void setSize(String size){ 
        this.size = size;
        switch(size){
            case "pequeña": price += 8; break;
            case "mediana": price += 10; break;
            case "grande": price += 12; break;
        }   
        
    }

    public void setDough(String dough){
        this.dough = dough; 
    }


    public void addIngredient(String ingredient, double cost){
        ingredients.add(ingredient);
        price += cost;
    }

    @Override
    public String toString(){
        return "Pizza" + size + ", masa" +dough +
        "\nIngredientes: " + String.join(",", ingredients) +
        "\nPrecio Total: $" + price;
    }
}

class PizzaBuilder {
    private Pizza pizza;

    public PizzaBuilder() {
        pizza = new Pizza();
    }

    public PizzaBuilder setSize(String size) {
        pizza.setSize(size);
        return this;
    }

    public PizzaBuilder setDough(String dough) {
        pizza.setDough(dough);
        return this;
    }

    public PizzaBuilder addIngredient(String ingredient, double cost) {
        pizza.addIngredient(ingredient, cost);
        return this;
    }

    public Pizza build() {
        return pizza;
    }
}

public class Builder {
    public static void main(String[] args) {
        PizzaBuilder builder = new PizzaBuilder();
        Pizza pizza = builder.setSize("mediana")
                            .setDough("fina")
                            .addIngredient("queso extra", 2)
                            .addIngredient("pepperoni", 3)
                            .build();

        System.out.println(pizza);
    }
}